File: day-01/solution.kt
import java.io.File

class Elf(private val inventory: List<Int>) {
    fun totalCalories() = inventory.sum()
}

fun parse(): MutableList<Elf> {
    var elf: MutableList<Int> = mutableListOf()
    val elves: MutableList<Elf> = mutableListOf()

    File("./input.txt").forEachLine { line ->
        run {
            if (line == "") {
                elves.add(Elf(elf))
                elf = mutableListOf()
            } else elf.add(line.toInt())
        }
    }
    return elves
}

fun main() {
    val elves = parse().map { elf -> elf.totalCalories() }
    val top = elves.max()
    println("Top Elf: $top")

    val top3 = elves.sortedDescending().take(3).sum()
    println("Top 3 Elves combined: $top3")
}

File: day-01/solution.cpp
#include <string>
#include <fstream>
#include <iostream>
#include <vector>
#include <algorithm>

std::vector<int> parse() {
    std::ifstream infile("input.txt");
    std::string line;
    std::vector<int> elves;
    int elf = 0;
    while(std::getline(infile, line)) {
        if (line.empty()) {
            elves.push_back(elf);
            elf = 0;
        } else {
            elf += std::stoi(line);
        }
    }
    return elves;
}

int main() {
    std::vector<int> elves = parse();
    std::sort(elves.begin(), elves.end());

    int top = elves.back();
    int top3 = 0;
    for (int i = 0; i < 3; i++) {
        top3 += elves.back();
        elves.pop_back();
    }

    std::cout << "Top: " << top << std::endl;
    std::cout << "Top 3: " << top3 << std::endl;
    return 0;
}